smarts/core/controllers/trajectory_interpolation_controller.py

```diff
@@ -0,0 +1,122 @@
+        np.isnan(np.sum(ndarray))
+
+    @staticmethod
+    def interpolate(ms0: np.ndarray, ms1: np.ndarray, time: float) -> np.ndarray:
+        """Linear Interpolate between two vehicle motion state
+
+        Returns:
+            np.ndarray: New vehicle state between vehicle motion state ms0 and ms1
+        """
+
+        start_time = ms0[TIME_INDEX]
+        end_time = ms1[TIME_INDEX]
+        ratio = math.fabs((time - start_time) / (end_time - start_time))
+        assert end_time >= start_time and time >= start_time
+
+        np_motion_state = (1 - ratio) * ms0 + ratio * ms1
+        return np_motion_state
+
+    @staticmethod
+    def locate_motion_state(trajectory, time) -> np.ndarray:
+
+        if math.isinf(trajectory[TIME_INDEX][0]):
+            return trajectory[:, 0], trajectory[:, 0]
+
+        end_index = 0
+        for i, t in enumerate(trajectory[TIME_INDEX]):
+            if t > time:
+                end_index = i
+                break
+
+        if end_index <= 0:
+            raise RuntimeError(
+                "Next step time is %f, but cant be located at input traj" % time
+            )
+
+        return trajectory[:, end_index - 1], trajectory[:, end_index]
+
+    @classmethod
+    def perform_trajectory_interpolation(
+        cls, sim, agent_id, vehicle, trajectory: np.ndarray, controller_state,
+    ):
+        """Move vehicle with trajectory interpolation
+
+        If you want vehicle stop at a specific pose,
+        trajectory[TIME_INDEX][0] should be setted as infi
+
+        Args:
+            sim : reference of smarts instance
+            agent_id : agent who use this controller
+            vehicle : vehicle to be controlled
+            trajectory (np.ndarray): [description]
+            controller_state : inner state of controller
+
+        Raises:
+            ValueError: Error when input trajectory has nan field.
+            RuntimeError: Error when input trajectory is empty
+        """
+        assert isinstance(vehicle.chassis, AckermannChassis)
+        if len(trajectory[TIME_INDEX]) <= 0:
+            raise RuntimeError("Input trajectory is empty!")
+        if cls.ndarray_has_nan(trajectory):
+            raise ValueError("Has nan in trajectory")
+
+        ms0, ms1 = cls.locate_motion_state(trajectory, sim.timestep_sec)
+
+        speed = 0.0
+        pose = []
+        if math.isinf(ms0[TIME_INDEX]) or math.isinf(ms1[TIME_INDEX]):
+            center_position = ms0[X_INDEX : Y_INDEX + 1]
+            center_heading = Heading(ms0[THETA_INDEX])
+            pose = Pose.from_center(center_position, center_heading)
+            speed = 0.0
+        else:
+            ms = cls.interpolate(ms0, ms1, sim.timestep_sec)
+
+            center_position = ms[X_INDEX : Y_INDEX + 1]
+            center_heading = Heading(ms[THETA_INDEX])
+            pose = Pose.from_center(center_position, center_heading)
+            speed = ms[VEL_INDEX]
+
+        vehicle.set_pose(pose)
```

Comment:
This seems like a reasonable call but it will not work unless the vehicle chassis is a box model at current, except the Ackermann chassis is the standard for agents. This is a problem since the call in the Ackermann chassis goes down to `bullet_client.resetBasePositionAndOrientation()` which disrupts the bullet simulation and causes collision filtering artefacts.

@liamchzh we have some options here, we can force the Ackermann chassis `set_pose()` to work as expected or swap the vehicle chassis to box model when using a kinematic controller.

Reply:
Let's support a fix within this sprint.